k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java

```diff
@@ -177,9 +177,46 @@ private void fillContactDataFromQuery(String query, List<Recipient> recipients,
             Map<String, Recipient> recipientMap) {
 
         ContentResolver contentResolver = getContext().getContentResolver();
-
         query = "%" + query + "%";
+
+        final String[] PROJECTION_NICKNAME =
+                new String[]{ContactsContract.Data.CONTACT_ID, ContactsContract.CommonDataKinds.Nickname.NAME};
+        final int INDEX_CONTACT_ID = 0;
+        final int INDEX_NICKNAME = 1;
+
         Uri queryUri = ContactsContract.CommonDataKinds.Email.CONTENT_URI;
+        Uri queryUriForNickname = ContactsContract.Data.CONTENT_URI;
+
+        /**
+         * Look for contacts whose nickname contains given string, retrieve CONTACT_ID and nickname
+         * of them
+         */
+        Cursor cursorForNickname = contentResolver.query(queryUriForNickname,
+                PROJECTION_NICKNAME,
+                ContactsContract.CommonDataKinds.Nickname.NAME + " LIKE ? AND " +
+                        Data.MIMETYPE + " = '" +
+                        ContactsContract.CommonDataKinds.Nickname.CONTENT_ITEM_TYPE + "'",
+                new String[] { query },
+                null);
+
+        /**
+         * query email etc. according to CONTACT_ID, and specify the name of recipient
+         */
+        if (cursorForNickname != null && cursorForNickname.moveToFirst()) {
```

Comment:
And you could do create a method for this logic too.

Smaller methods are easier to read, the logic is more obvious and so less prone to error.